Add a `breakpoint add` command to fix the option-madness that is `breakpoint set`

jimingham opened the pull request:
Someone came up with a clever idea for a new breakpoint type, but we couldn't figure out how to add it in an ergonomic way because `breakpoint set` has used up all the short-option characters.  And even if they did find a way to add it, the help for `break set` is so confusing - because of the way it is implemented - that very few people would detect the addition.  

The basic problem is that `break set` distinguishes amongst the fundamental breakpoint types it offers by which "required option" you provide.  If you pass a `-a` you are setting an address breakpoint, if `-n`, `-F`, etc. a symbol name based breakpoint.  And so forth.  That is however pretty hard to discern from the option grouping printing from `help break set`.  `break set` also suffers from the problem that it uses common options in different ways depending on which "required" option is present, which makes documenting the various behaviors difficult.  And as we run out of single letters it makes extending it difficult to impossible.

This PR fixes that situation by adding a new command for adding breakpoints - `break add`.  The new command specifies the "breakpoint types" as subcommands of `break add` rather than distinguishing them by their being one of the "required" options the way `break set` does.  That both makes it much clearer what the breakpoint types actually are, and means that the option set can be dedicated to that particular breakpoint type, and so the help for each is less cluttered, and can be documented properly for each usage.

Instead of trying to parse the meaning of:

```
(lldb) help break set
Sets a breakpoint or set of breakpoints in the executable.

Syntax: breakpoint set <cmd-options>

Command Options Usage:
  breakpoint set [-DHd] -l <linenum> [-G <boolean>] [-C <command>] [-c <expr>] [-Y <source-language>] [-i <count>] [-o <boolean>] [-q <queue-name>] [-t <thread-id>] [-x <thread-index>] [-T <thread-name>] [-R <address>] [-N <breakpoint-name>] [-u <column>] [-f <filename>] [-m <boolean>] [-s <shlib-name>] [-K <boolean>]
  breakpoint set [-DHd] -a <address-expression> [-G <boolean>] [-C <command>] [-c <expr>] [-Y <source-language>] [-i <count>] [-o <boolean>] [-q <queue-name>] [-t <thread-id>] [-x <thread-index>] [-T <thread-name>] [-N <breakpoint-name>] [-s <shlib-name>]
  breakpoint set [-DHd] -n <function-name> [-G <boolean>] [-C <command>] [-c <expr>] [-Y <source-language>] [-i <count>] [-o <boolean>] [-q <queue-name>] [-t <thread-id>] [-x <thread-index>] [-T <thread-name>] [-R <address>] [-N <breakpoint-name>] [-f <filename>] [-L <source-language>] [-s <shlib-name>] [-K <boolean>]
  breakpoint set [-DHd] -F <fullname> [-G <boolean>] [-C <command>] [-c <expr>] [-Y <source-language>] [-i <count>] [-o <boolean>] [-q <queue-name>] [-t <thread-id>] [-x <thread-index>] [-T <thread-name>] [-R <address>] [-N <breakpoint-name>] [-f <filename>] [-L <source-language>] [-s <shlib-name>] [-K <boolean>]
  breakpoint set [-DHd] -S <selector> [-G <boolean>] [-C <command>] [-c <expr>] [-Y <source-language>] [-i <count>] [-o <boolean>] [-q <queue-name>] [-t <thread-id>] [-x <thread-index>] [-T <thread-name>] [-R <address>] [-N <breakpoint-name>] [-f <filename>] [-L <source-language>] [-s <shlib-name>] [-K <boolean>]
  breakpoint set [-DHd] -M <method> [-G <boolean>] [-C <command>] [-c <expr>] [-Y <source-language>] [-i <count>] [-o <boolean>] [-q <queue-name>] [-t <thread-id>] [-x <thread-index>] [-T <thread-name>] [-R <address>] [-N <breakpoint-name>] [-f <filename>] [-L <source-language>] [-s <shlib-name>] [-K <boolean>]
  breakpoint set [-DHd] -r <regular-expression> [-G <boolean>] [-C <command>] [-c <expr>] [-Y <source-language>] [-i <count>] [-o <boolean>] [-q <queue-name>] [-t <thread-id>] [-x <thread-index>] [-T <thread-name>] [-R <address>] [-N <breakpoint-name>] [-f <filename>] [-L <source-language>] [-s <shlib-name>] [-K <boolean>]
  breakpoint set [-DHd] -b <function-name> [-G <boolean>] [-C <command>] [-c <expr>] [-Y <source-language>] [-i <count>] [-o <boolean>] [-q <queue-name>] [-t <thread-id>] [-x <thread-index>] [-T <thread-name>] [-R <address>] [-N <breakpoint-name>] [-f <filename>] [-L <source-language>] [-s <shlib-name>] [-K <boolean>]
  breakpoint set [-ADHd] -p <regular-expression> [-G <boolean>] [-C <command>] [-c <expr>] [-Y <source-language>] [-i <count>] [-o <boolean>] [-q <queue-name>] [-t <thread-id>] [-x <thread-index>] [-T <thread-name>] [-N <breakpoint-name>] [-f <filename>] [-m <boolean>] [-s <shlib-name>] [-X <function-name>]
  breakpoint set [-DHd] -E <source-language> [-G <boolean>] [-C <command>] [-c <expr>] [-Y <source-language>] [-i <count>] [-o <boolean>] [-q <queue-name>] [-t <thread-id>] [-x <thread-index>] [-T <thread-name>] [-N <breakpoint-name>] [-h <boolean>] [-w <boolean>]
  breakpoint set [-DHd] -P <python-class> [-k <none>] [-v <none>] [-G <boolean>] [-C <command>] [-c <expr>] [-Y <source-language>] [-i <count>] [-o <boolean>] [-q <queue-name>] [-t <thread-id>] [-x <thread-index>] [-T <thread-name>] [-N <breakpoint-name>] [-f <filename>] [-s <shlib-name>]
  breakpoint set [-DHd] -y <linespec> [-G <boolean>] [-C <command>] [-c <expr>] [-Y <source-language>] [-i <count>] [-o <boolean>] [-q <queue-name>] [-t <thread-id>] [-x <thread-index>] [-T <thread-name>] [-R <address>] [-N <breakpoint-name>] [-m <boolean>] [-s <shlib-name>] [-K <boolean>]

```

We instead offer:

```
(lldb) help break add
Commands to add breakpoints of various types

Syntax: breakpoint add

Access the breakpoint search kernels built into lldb.  Along with specifying the
search kernel, each breakpoint add operation can specify a common set of 
"reaction" options for each breakpoint.  The reaction options can also be
modified after breakpoint creation using the "breakpoint modify" command.       
        

The following subcommands are supported:

      address   -- Add breakpoints by raw address
      exception -- Add breakpoints on language exceptions.  If no language is specified, break on exceptions for all supported languages
      file      -- Add breakpoints on lines in specified source files
      name      -- Add breakpoints matching function or symbol names
      pattern   -- Add breakpoints matching patterns in the source text  Expects 'raw' input (see 'help raw-input'.)
      scripted  -- Add breakpoints using a scripted breakpoint resolver.

For more help on any particular subcommand, type 'help <command> <subcommand>'.

```

The individual subcommand helps are also easier to read.  They are still a little too verbose because they all repeat the options for the `reactions`.  A general fix for our help system would be when a command incorporates an OptionGroup whole into the command options, help would show the option group name - which you could separately look up.  But even without that:

```
(lldb) help br a a
Add breakpoints by raw address

Syntax: breakpoint add address <cmd-options> <address> [<address> [...]]

Command Options Usage:
  breakpoint add address [-DHde] [-G <boolean>] [-C <command>] [-c <expr>] [-Y <source-language>] [-i <count>] [-o <boolean>] [-q <queue-name>] [-t <thread-id>] [-x <thread-index>] [-T <thread-name>] [-N <breakpoint-name>] [-s <shlib-name>] <address> [<address> [...]]

       -C <command> ( --command <command> )
            A command to run when the breakpoint is hit, can be provided more than once, the commands will be run in left-to-right order.

       -D ( --dummy-breakpoints )
            Act on Dummy breakpoints - i.e. breakpoints set before a file is provided, which prime new targets.

       -G <boolean> ( --auto-continue <boolean> )
            The breakpoint will auto-continue after running its commands.

       -H ( --hardware )
            Require the breakpoint to use hardware breakpoints.

       -N <breakpoint-name> ( --breakpoint-name <breakpoint-name> )
            Adds this name to the list of names for this breakpoint.  Can be specified more than once.

       -T <thread-name> ( --thread-name <thread-name> )
            The breakpoint stops only for the thread whose thread name matches this argument.

       -Y <source-language> ( --condition-language <source-language> )
            Specifies the Language to use when executing the breakpoint's condition expression.

       -c <expr> ( --condition <expr> )
            The breakpoint stops only if this condition expression evaluates to true.

       -d ( --disable )
            Disable the breakpoint.

       -e ( --enable )
            Enable the breakpoint.

       -i <count> ( --ignore-count <count> )
            Set the number of times this breakpoint is skipped before stopping.

       -o <boolean> ( --one-shot <boolean> )
            The breakpoint is deleted the first time it stop causes a stop.

       -q <queue-name> ( --queue-name <queue-name> )
            The breakpoint stops only for threads in the queue whose name is given by this argument.

       -s <shlib-name> ( --shlib <shlib-name> )
            Set the breakpoint at an address relative to sections in this shared library.

       -t <thread-id> ( --thread-id <thread-id> )
            The breakpoint stops only for the thread whose TID matches this argument.  The token 'current' resolves to the current thread's ID.

       -x <thread-index> ( --thread-index <thread-index> )
            The breakpoint stops only for the thread whose index matches this argument.
     
     This command takes options and free-form arguments.  If your arguments resemble option specifiers (i.e., they start with a - or --), you must use ' --
     ' between the end of the command options and the beginning of the arguments.

```

is pretty readable.

jimingham committed:
breakpoint add - all subcommands implemented.  Testsuite clean but no
new tests.

jimingham commented:
IIRC there's some formal way to mark PR's as draft's, but I didn't see where it was in the UI.

Anyway, I think that the code here is what I want, but the testing strategy needs to be fleshed out.  So far I haven't add any new capabilities to the `break add` commands that aren't in `break set`.  I tested it by adding the ability to switch the `lldbutils.run_break_set_...` routines, in the draft I hard coded it to use add.  That actually does a pretty good job of exercising the new code, because that's how we wrote most of the `break set` tests in the API tests.

But we don't want to not test `break set` and I don't think we want this to make two test variants.  So we still need to refine the testing strategy for this patch.

jimingham commented:
If we decide to go this way, I will also - probably as a follow-on - switch the __regex_breakpoint command over to use `break add`, though I haven't done that in this patch yet.

jimingham committed:
Merge branch 'main' into breakpoint-add

jimingham committed:
formatting.

jimingham committed:
Merge branch 'main' into breakpoint-add

jimingham committed:
Fix a test that was comparing against the old breakpoint subcommand list.

jimingham committed:
Merge branch 'main' into breakpoint-add

jimingham committed:
Merge branch 'main' into breakpoint-add

jimingham committed:
Merge branch 'main' into breakpoint-add

JDevlieghere commented on lldb/source/Commands/CommandObjectBreakpoint.cpp:
```diff
@@ -200,41 +200,1341 @@ class lldb_private::BreakpointOptionGroup : public OptionGroup {
+
+  bp_sp->GetOptions().CopyOverSetOptions(bp_opts.GetBreakpointOptions());
+  Target &target = bp_sp->GetTarget();
+  if (!bp_names.empty()) {
```
This seems redundant: we don't do anything special compared to not iterating to the list. Saving the stack-allocated Status doesn't offset the extra level of indentation.

JDevlieghere commented on lldb/source/Commands/CommandObjectBreakpoint.cpp:
```diff
@@ -200,41 +200,1341 @@ class lldb_private::BreakpointOptionGroup : public OptionGroup {
+    for (auto name : bp_names) {
+      target.AddNameToBreakpoint(bp_sp, name.c_str(), name_error);
+      if (name_error.Fail()) {
+        result.AppendErrorWithFormat("Invalid breakpoint name: %s",
```
Errors should start with lowercase letters (https://llvm.org/docs/CodingStandards.html#error-and-warning-messages)

JDevlieghere commented on lldb/source/Commands/CommandObjectBreakpoint.cpp:
```diff
@@ -200,41 +200,1341 @@ class lldb_private::BreakpointOptionGroup : public OptionGroup {
+  // First use the Source Manager's default file. Then use the current stack
+  // frame's file.
+  if (!exe_ctx.HasTargetScope()) {
+    error_msg = "Can't get a default file with no target.";
```
```suggestion
    error_msg = "can't get a default file with no target.";
```

JDevlieghere reviewed (approved):
LGTM, there's really not that much actual new code to review.  The code duplication is unfortunate but I don't think that should stand in the way of a more ergonomic command. Could we add a big warning at the top of the old and new file to remind folks to at least try to keep the two in sync for now?

I left some comments for improvements, though I realize this is all pre-existing code. We would be better off addressing it for both the old and new implementation in a follow up PR.

jimingham commented:
> LGTM, there's really not that much actual new code to review. The code duplication is unfortunate but I don't think that should stand in the way of a more ergonomic command. Could we add a big warning at the top of the old and new file to remind folks to at least try to keep the two in sync for now?

I would say "no more changes allowed to `break set` except mutatis mutandis for API changes."  I don't want to add new features in parallel for the two because that would severely limit the improvements we could make to `break add` - one of the big reasons for moving over to `break add` is that we can make improvements that were impossible with the older structure.  And if we're fixing bugs, they are for the most part going to be in code either one of these commands calls, not in the command parsing and dispatch.

So for the most part I'd like to declare `break set` frozen except for bugs, and have all future work and improvements done to `break add`.

> 
> I left some comments for improvements, though I realize this is all pre-existing code. We would be better off addressing it for both the old and new implementation in a follow up PR.

jimingham committed:
Merge branch 'main' into breakpoint-add

jimingham committed:
Merge branch 'main' into breakpoint-add

jimingham committed:
Merge branch 'main' into breakpoint-add

jimingham commented:
Okay, I have to do the merge conflict resolution by hand.  I haven't had much luck with using the GitHub UI, it always seems to mangle the changes.

medismailben commented on lldb/source/Commands/CommandObjectBreakpoint.cpp:
```diff
@@ -200,41 +200,1341 @@ class lldb_private::BreakpointOptionGroup : public OptionGroup {
+        output_stream.Printf("Breakpoint set in dummy target, will get copied "
+                             "into future targets.\n");
+      else {
+        // Don't print out this warning for exception breakpoints.  They can
```
nit: extra space
```suggestion
        // Don't print out this warning for exception breakpoints. They can
```

jimingham replied:
There are always two spaces after a period.

medismailben commented on lldb/source/Commands/CommandObjectBreakpoint.cpp:
```diff
@@ -200,41 +200,1341 @@ class lldb_private::BreakpointOptionGroup : public OptionGroup {
+        m_python_class_options.GetStructuredData(), &error);
+    if (error.Fail()) {
+      result.AppendErrorWithFormat(
+          "Error setting extra exception arguments: %s", error.AsCString());
```
start the first sentence with a lowercase letter

medismailben commented on lldb/source/Commands/CommandObjectBreakpoint.cpp:
```diff
@@ -200,41 +200,1341 @@ class lldb_private::BreakpointOptionGroup : public OptionGroup {
+Access the breakpoint search kernels built into lldb.  Along with specifying the
+search kernel, each breakpoint add operation can specify a common set of 
+"reaction" options for each breakpoint.  The reaction options can also be
+modified after breakpoint creation using the "breakpoint modify" command.       
```
extra spaces ? Is it `clang-format` who put these ?

```suggestion
Access the breakpoint search kernels built into lldb. Along with specifying the
search kernel, each breakpoint add operation can specify a common set of 
"reaction" options for each breakpoint. The reaction options can also be
modified after breakpoint creation using the "breakpoint modify" command.       
```

jimingham replied:
In real English (as opposed to what you type into chat fields) there are always two spaces after a period.

medismailben commented on lldb/source/Interpreter/Options.cpp:
```diff
@@ -1410,7 +1410,9 @@ llvm::Error lldb_private::CreateOptionParsingError(
   std::string buffer;
   llvm::raw_string_ostream stream(buffer);
-  stream << "Invalid value ('" << option_arg << "') for -" << short_option;
+  stream << "Invalid value ('" << option_arg << "')";
```
"start the first sentence with a lowercase letter"
```suggestion
  stream << "invalid value ('" << option_arg << "')";
```

medismailben reviewed (approved):
LGTM with comments.

jimingham committed:
Rework the change to GetDefaultFile since I need to use it in two
CommandObjectBreakpoint classes.
React to Ismail's review suggestions.

jimingham committed:
Merge branch 'main' into breakpoint-add

jimingham committed:
formatting

jimingham committed:
Another couple places with an error check for "Invalid value"

I changed that to "invalid" in response to review comments, and
that had test fallout.

jimingham committed:
Trying to figure out why the "symbols-on-demand" test isn't working
only for "dwarf" and only on one of the Linux bots...

jimingham committed:
See if providing the file will make the on-demand do its thing.

jimingham committed:
Merge branch 'main' into breakpoint-add

jimingham committed:
formatting

jimingham committed:
Merge branch 'main' into breakpoint-add
